doc/source/revent.rst

```diff
@@ -210,3 +210,99 @@ https://www.kernel.org/doc/Documentation/input/event-codes.txt
+
+
+Using state detection with revent
+=================================
+
+State detection can be used to verify that a workload is executing as expected.
+This utility, if enabled, and if state definitions are available for the
+particular workload, takes a screenshot after the setup and the run revent
+sequence, matches the screenshot to a state and compares with the expected
+state. A WorkloadError is raised if an unexpected state is encountered.
+
+To enable state detection, make sure a valid state definition file and
+templates exist for your workload and set the check_states parameter to True.
+
+State definition directory
+--------------------------
+
+State and phase definitions should be placed in a directory of the following
+structure inside the dependencies directory of each workload (along with
+revent files etc):
+
+   dependencies/
+      <workload_name>/
+         state_definitions/
+            definition.yaml
+            templates/
+               <oneTemplate>.png
+               <anotherTemplate>.png
+               ...
+
+definition.yaml file
+--------------------
+
+This defines each state of the workload and lists which templates are expected
+to be found and how many are required to be detected for a conclusive match. It
+also defines the expected state in each workload phase where a state detection
+is run (currently those are setupComplete and runComplete).
+
+Templates are picture elements to be matched in a screenshot. Each template
+mentioned in the definition file should be placed as a file with the same name
+and a .png extension inside the templates folder. Creating template png files
+is as simple as taking a screenshot of the workload in a given state, cropping
+out the relevant templates (eg. a button, label or other unique element that is
+present in that state) and storing them in PNG format.
+
+Please see the definition file for Angry Birds below as an example to
+understand the format:
+
+.. code-block:: python
+
+    workloadName: angrybirds
+
+    workloadStates:
+      - stateName: titleScreen
+        templates:
+          - playButton
+          - logo
+        matches: 2
+      - stateName: worldSelection
+        templates:
+          - firstWorldThumb
+          - secondWorldThumb
+          - thirdWorldThumb
+          - fourthWorldThumb
+        matches: 3
+      - stateName: levelSelection
+        templates:
+          - lockedLevel
+          - firstLevel
+        matches: 2
+      - stateName: gameplay
+        templates:
+          - pauseButton
+          - scoreLabelText
+        matches: 2
+      - stateName: pauseScreen
+        templates:
+          - replayButton
+          - menuButton
+          - resumeButton
+          - helpButton
+        matches: 4
+      - stateName: levelClearedScreen
+        templates:
+          - levelClearedText
+          - menuButton
+          - replayButton
+          - fastForwardButton
+        matches: 4
+
+    workloadPhases:
```

Comment:
While only those two states are required, more can be defined so that the current state of the workload can be indicated even in a mismatch (for example, "Unexpected state: pauseScreen instead of levelClearScreen"). I defined most of those states for demo purposes in Angry Birds.